File: src/main/java/io/gluth/pagespace/layout/ForceDirectedLayout.java
package io.gluth.pagespace.layout;

import io.gluth.pagespace.domain.Page;
import io.gluth.pagespace.domain.PageGraph;

import java.util.*;

/**
 * 3-D Fruchterman-Reingold force-directed layout with adaptive settling.
 *
 * Oscillation detector: when a node's velocity vector reverses direction
 * (dot product with the previous velocity is negative), a moderate extra-
 * damping impulse is applied and a still-counter accumulates.  Slow
 * residual motion increments the counter one step at a time; fast motion
 * without reversal decrements it slowly (so brief excursions don't reset
 * all progress).  When the counter reaches SETTLE_STEPS the node's velocity
 * is zeroed and subsequent force updates are skipped — the node is frozen.
 *
 * Navigation events (setPinnedPage / syncWithGraph) reset all counters so
 * every node re-animates and re-settles around the new focus.
 */
public class ForceDirectedLayout {

    private static final double BASE_DAMPING     = 0.75;
    private static final double FORCE_SCALE      = 0.08;
    private static final double MAX_SPEED        = 60.0;
    /** Extra-damping multiplier applied on each velocity reversal. */
    private static final double REVERSAL_DAMP    = 0.45;
    /** Still-counter credits added per reversal. */
    private static final int    REVERSAL_CREDIT  = 8;
    /** Speed below which the still-counter increments each step. */
    private static final double SETTLE_THRESHOLD = 4.0;
    /** Counter decrement per step when moving fast without reversal. */
    private static final int    FAST_DECAY       = 2;
    /** Counter value at which the node is considered fully settled. */
    private static final int    SETTLE_STEPS     = 40;

    private static final double BOUNDARY_MARGIN   = 0.12;
    private static final double BOUNDARY_STRENGTH = 50.0;
    /** Steps over which the focus node travels from its old position to centre. */
    private static final int    TRANSITION_STEPS  = 35;

    private final PageGraph graph;
    private final double width;
    private final double height;
    private final double depth;
    private final Random random;

    private final Map<Page, NodePosition> positions   = new LinkedHashMap<>();
    private final Map<Page, double[]>     velocities  = new LinkedHashMap<>();
    private final Map<Page, Integer>      stillCounts = new LinkedHashMap<>();

    private Page   pinnedPage;
    private double transStartX, transStartY, transStartZ;
    private int    transStep = TRANSITION_STEPS; // starts "done" so first pinning transitions

    public ForceDirectedLayout(PageGraph graph, double width, double height) {
        this(graph, width, height, System.nanoTime());
    }

    public ForceDirectedLayout(PageGraph graph, double width, double height, long seed) {
        this.graph  = graph;
        this.width  = width;
        this.height = height;
        this.depth  = Math.min(width, height);
        this.random = new Random(seed);
        initPositions();
    }

    private void initPositions() {
        for (Page page : graph.pages()) {
            if (!positions.containsKey(page)) {
                positions.put(page,    spawnPosition(page));
                velocities.put(page,   new double[3]);
                stillCounts.put(page,  0);
            }
        }
    }

    private NodePosition spawnPosition(Page page) {
        List<NodePosition> nb = new ArrayList<>();
        for (Page n : graph.linksFrom(page)) { NodePosition np = positions.get(n); if (np != null) nb.add(np); }
        for (Page n : graph.linksTo(page))   { NodePosition np = positions.get(n); if (np != null) nb.add(np); }
        if (!nb.isEmpty()) {
            double sx = 0, sy = 0, sz = 0;
            for (NodePosition np : nb) { sx += np.x(); sy += np.y(); sz += np.z(); }
            double cnt = nb.size(), j = Math.min(width, height) * 0.18;
            return new NodePosition(
                clamp(sx/cnt + (random.nextDouble()-0.5)*j,        0, width),
                clamp(sy/cnt + (random.nextDouble()-0.5)*j,        0, height),
                clamp(sz/cnt + (random.nextDouble()-0.5)*j*0.4,   0, depth));
        }
        double xm = width*BOUNDARY_MARGIN, ym = height*BOUNDARY_MARGIN, zm = depth*BOUNDARY_MARGIN;
        return new NodePosition(
            xm + random.nextDouble()*(width  - 2*xm),
            ym + random.nextDouble()*(height - 2*ym),
            zm + random.nextDouble()*(depth  - 2*zm));
    }

    public void setPinnedPage(Page page) {
        this.pinnedPage = page;
        NodePosition pos = positions.get(page);
        if (pos != null) {
            transStartX = pos.x();
            transStartY = pos.y();
            transStartZ = pos.z();
        } else {
            transStartX = width / 2.0;
            transStartY = height / 2.0;
            transStartZ = 0.0;
        }
        transStep = 0;
        wakeAll();
    }

    public void syncWithGraph() {
        boolean added = false;
        for (Page page : graph.pages()) {
            if (!positions.containsKey(page)) {
                positions.put(page,   spawnPosition(page));
                velocities.put(page,  new double[3]);
                stillCounts.put(page, 0);
                added = true;
            }
        }
        if (added) wakeAll();
    }

    /** Reset all still-counters so every node re-animates after a navigation event. */
    private void wakeAll() {
        for (Page p : stillCounts.keySet()) stillCounts.put(p, 0);
        for (double[] v : velocities.values()) { v[0] = 0; v[1] = 0; v[2] = 0; }
    }

    public void step() {
        List<Page> pageList = new ArrayList<>(positions.keySet());
        int n = pageList.size();
        if (n == 0) return;

        double k  = Math.cbrt((width * height * depth) / Math.max(n, 1));
        double xm = width*BOUNDARY_MARGIN, ym = height*BOUNDARY_MARGIN, zm = depth*BOUNDARY_MARGIN;

        Map<Page, double[]> forces = new HashMap<>();
        for (Page p : pageList) forces.put(p, new double[3]);

        // Repulsion between all pairs
        for (int i = 0; i < n; i++) {
            Page u = pageList.get(i); NodePosition pu = positions.get(u);
            for (int j = i + 1; j < n; j++) {
                Page v = pageList.get(j); NodePosition pv = positions.get(v);
                double dx=pu.x()-pv.x(), dy=pu.y()-pv.y(), dz=pu.z()-pv.z();
                double dist=Math.max(Math.sqrt(dx*dx+dy*dy+dz*dz),0.001), rep=(k*k)/dist;
                double[] fu=forces.get(u), fv=forces.get(v);
                fu[0]+=(dx/dist)*rep; fu[1]+=(dy/dist)*rep; fu[2]+=(dz/dist)*rep;
                fv[0]-=(dx/dist)*rep; fv[1]-=(dy/dist)*rep; fv[2]-=(dz/dist)*rep;
            }
        }

        // Attraction along edges weighted by shared-link count
        for (Page u : pageList) {
            for (Page v : graph.linksFrom(u)) {
                if (!positions.containsKey(v)) continue;
                NodePosition pu=positions.get(u), pv=positions.get(v);
                double dx=pu.x()-pv.x(), dy=pu.y()-pv.y(), dz=pu.z()-pv.z();
                double dist=Math.max(Math.sqrt(dx*dx+dy*dy+dz*dz),0.001);
                double att=(dist*dist)/k*(1.0+graph.sharedLinkCount(u,v));
                double[] fu=forces.get(u), fv=forces.get(v);
                fu[0]-=(dx/dist)*att; fu[1]-=(dy/dist)*att; fu[2]-=(dz/dist)*att;
                fv[0]+=(dx/dist)*att; fv[1]+=(dy/dist)*att; fv[2]+=(dz/dist)*att;
            }
        }

        // Quadratic boundary repulsion
        for (Page p : pageList) {
            if (p.equals(pinnedPage)) continue;
            NodePosition pos=positions.get(p); double[] f=forces.get(p);
            if (pos.x()<xm)          {double t=(xm-pos.x())/xm;          f[0]+=k*t*t*BOUNDARY_STRENGTH;}
            if (pos.x()>width-xm)    {double t=(pos.x()-(width-xm))/xm;  f[0]-=k*t*t*BOUNDARY_STRENGTH;}
            if (pos.y()<ym)          {double t=(ym-pos.y())/ym;          f[1]+=k*t*t*BOUNDARY_STRENGTH;}
            if (pos.y()>height-ym)   {double t=(pos.y()-(height-ym))/ym; f[1]-=k*t*t*BOUNDARY_STRENGTH;}
            if (pos.z()<zm)          {double t=(zm-pos.z())/zm;          f[2]+=k*t*t*BOUNDARY_STRENGTH;}
            if (pos.z()>depth-zm)    {double t=(pos.z()-(depth-zm))/zm;  f[2]-=k*t*t*BOUNDARY_STRENGTH;}
        }

        // Integrate: skip frozen nodes entirely so applied forces don't re-wake them.
        for (Page p : pageList) {
            if (p.equals(pinnedPage)) continue;

            int count = stillCounts.getOrDefault(p, 0);
            if (count >= SETTLE_STEPS) continue;  // frozen — don't touch velocity

            double[] f=forces.get(p), vel=velocities.get(p);
            NodePosition pos=positions.get(p);

            double pv0=vel[0], pv1=vel[1], pv2=vel[2];

            vel[0]=(vel[0]+f[0]*FORCE_SCALE)*BASE_DAMPING;
            vel[1]=(vel[1]+f[1]*FORCE_SCALE)*BASE_DAMPING;
            vel[2]=(vel[2]+f[2]*FORCE_SCALE)*BASE_DAMPING;

            double speed=Math.sqrt(vel[0]*vel[0]+vel[1]*vel[1]+vel[2]*vel[2]);
            if (speed>MAX_SPEED) {
                double s=MAX_SPEED/speed; vel[0]*=s; vel[1]*=s; vel[2]*=s; speed=MAX_SPEED;
            }

            double dot = pv0*vel[0]+pv1*vel[1]+pv2*vel[2];
            if (dot < 0) {
                // Velocity reversal — approaching equilibrium; apply extra damping
                vel[0]*=REVERSAL_DAMP; vel[1]*=REVERSAL_DAMP; vel[2]*=REVERSAL_DAMP;
                speed*=REVERSAL_DAMP;
                count = Math.min(count+REVERSAL_CREDIT, SETTLE_STEPS);
            } else if (speed < SETTLE_THRESHOLD) {
                count = Math.min(count+1, SETTLE_STEPS);
            } else {
                // Moving fast without reversal: decay counter slowly
                count = Math.max(0, count-FAST_DECAY);
            }
            stillCounts.put(p, count);

            if (count >= SETTLE_STEPS) {
                vel[0]=0; vel[1]=0; vel[2]=0;
            } else if (count > 0) {
                double suppress = 1.0-(double)count/SETTLE_STEPS;
                vel[0]*=suppress; vel[1]*=suppress; vel[2]*=suppress;
            }

            pos.setX(clamp(pos.x()+vel[0], 0, width));
            pos.setY(clamp(pos.y()+vel[1], 0, height));
            pos.setZ(clamp(pos.z()+vel[2], 0, depth));
        }

        // Animate pinned node toward centre using a smoothstep curve
        if (pinnedPage != null && positions.containsKey(pinnedPage)) {
            NodePosition pp = positions.get(pinnedPage);
            if (transStep < TRANSITION_STEPS) {
                transStep++;
                double t  = (double) transStep / TRANSITION_STEPS;
                double t2 = t * t * (3 - 2 * t);           // smoothstep
                pp.setX(transStartX + (width  / 2.0 - transStartX) * t2);
                pp.setY(transStartY + (height / 2.0 - transStartY) * t2);
                pp.setZ(transStartZ + (0.0          - transStartZ) * t2);
            } else {
                pp.setX(width / 2.0);
                pp.setY(height / 2.0);
                pp.setZ(0.0);
            }
        }
    }

    public Map<Page, NodePosition> positions()  { return Collections.unmodifiableMap(positions); }
    public PageGraph graph()       { return graph; }
    public double width()          { return width; }
    public double height()         { return height; }
    public double depth()          { return depth; }
    public int    transitionSteps(){ return TRANSITION_STEPS; }

    private static double clamp(double v, double lo, double hi) {
        return Math.max(lo, Math.min(hi, v));
    }
}

File: src/main/kotlin/io/gluth/pagespace/layout/ForceDirectedLayout.kt
package io.gluth.pagespace.layout

import io.gluth.pagespace.domain.Page
import io.gluth.pagespace.domain.PageGraph
import java.util.Collections
import kotlin.math.*

class ForceDirectedLayout(
    val graph: PageGraph,
    val width: Double,
    val height: Double,
    seed: Long = System.nanoTime()
) {

    companion object {
        private const val BASE_DAMPING     = 0.75
        private const val FORCE_SCALE      = 0.08
        private const val MAX_SPEED        = 60.0
        private const val REVERSAL_DAMP    = 0.45
        private const val REVERSAL_CREDIT  = 8
        private const val SETTLE_THRESHOLD = 4.0
        private const val FAST_DECAY       = 2
        private const val SETTLE_STEPS     = 40
        private const val SPHERE_MARGIN     = 0.12
        private const val BOUNDARY_STRENGTH = 50.0
        private const val TRANSITION_STEPS = 35
        private const val MAX_COMPUTE_ITERATIONS = 500
    }

    enum class Phase { IDLE, ANIMATING }

    val depth: Double = min(width, height)
    val sphereRadius: Double = min(width, height) / 2.0
    private val cx = width / 2.0
    private val cy = height / 2.0
    private val cz = depth / 2.0
    val transitionSteps: Int = TRANSITION_STEPS

    private val random = java.util.Random(seed)

    private val _positions:  MutableMap<Page, NodePosition> = LinkedHashMap()
    private val velocities:  MutableMap<Page, DoubleArray>  = LinkedHashMap()
    private val stillCounts: MutableMap<Page, Int>          = LinkedHashMap()

    private var pinnedPage: Page? = null

    private var phase: Phase = Phase.IDLE
    private val animStart:  MutableMap<Page, NodePosition> = LinkedHashMap()
    private val animTarget: MutableMap<Page, NodePosition> = LinkedHashMap()
    private var animStep: Int = 0

    val positions: Map<Page, NodePosition>
        get() = Collections.unmodifiableMap(_positions)

    val isSettled: Boolean
        get() = isConverged()

    init {
        initPositions()
    }

    private fun initPositions() {
        for (page in graph.pages()) {
            if (!_positions.containsKey(page)) {
                _positions[page]   = spawnPosition(page)
                velocities[page]   = DoubleArray(3)
                stillCounts[page]  = 0
            }
        }
    }

    private fun spawnPosition(page: Page): NodePosition {
        val nb = mutableListOf<NodePosition>()
        for (n in graph.linksFrom(page)) { _positions[n]?.let { nb.add(it) } }
        for (n in graph.linksTo(page))   { _positions[n]?.let { nb.add(it) } }
        if (nb.isNotEmpty()) {
            val sx = nb.sumOf { it.x }
            val sy = nb.sumOf { it.y }
            val sz = nb.sumOf { it.z }
            val cnt = nb.size.toDouble()
            val j = min(width, height) * 0.18
            return NodePosition(
                clamp(sx / cnt + (random.nextDouble() - 0.5) * j,       0.0, width),
                clamp(sy / cnt + (random.nextDouble() - 0.5) * j,       0.0, height),
                clamp(sz / cnt + (random.nextDouble() - 0.5) * j * 0.4, 0.0, depth))
        }
        // Random point inside sphere (radius * 0.8 to keep away from boundary)
        val r   = sphereRadius * 0.8 * cbrt(random.nextDouble())
        val theta = random.nextDouble() * 2 * Math.PI
        val phi   = Math.acos(2 * random.nextDouble() - 1)
        return NodePosition(
            cx + r * Math.sin(phi) * Math.cos(theta),
            cy + r * Math.sin(phi) * Math.sin(theta),
            cz + r * Math.cos(phi))
    }

    fun setPinnedPage(page: Page) {
        pinnedPage = page
    }

    fun syncWithGraph() {
        var added = false
        for (page in graph.pages()) {
            if (!_positions.containsKey(page)) {
                _positions[page]   = spawnPosition(page)
                velocities[page]   = DoubleArray(3)
                stillCounts[page]  = 0
                added = true
            }
        }
        if (added) wakeAll()
    }

    private fun wakeAll() {
        for (p in stillCounts.keys) stillCounts[p] = 0
        for (v in velocities.values) { v[0] = 0.0; v[1] = 0.0; v[2] = 0.0 }
    }

    fun computeEquilibrium() {
        // 1. Snapshot current positions as animation start
        animStart.clear()
        for ((page, pos) in _positions) {
            animStart[page] = NodePosition(pos.x, pos.y, pos.z)
        }

        // 2. Wake all nodes and run force simulation to convergence
        wakeAll()
        for (i in 0 until MAX_COMPUTE_ITERATIONS) {
            stepForces()
            if (isConverged()) break
        }

        // 3. Snapshot converged positions as animation targets
        animTarget.clear()
        for ((page, pos) in _positions) {
            animTarget[page] = NodePosition(pos.x, pos.y, pos.z)
        }

        // 4. Restore positions to animation start
        for ((page, startPos) in animStart) {
            val pos = _positions[page] ?: continue
            pos.x = startPos.x
            pos.y = startPos.y
            pos.z = startPos.z
        }

        // 5. Begin animation phase
        animStep = 0
        phase = Phase.ANIMATING
    }

    fun completeAnimation() {
        for ((page, target) in animTarget) {
            val pos = _positions[page] ?: continue
            pos.x = target.x
            pos.y = target.y
            pos.z = target.z
        }
        phase = Phase.IDLE
        animStart.clear()
        animTarget.clear()
    }

    private fun isConverged(): Boolean {
        if (_positions.isEmpty()) return false
        for ((page, count) in stillCounts) {
            if (page == pinnedPage) continue
            if (count < SETTLE_STEPS) return false
        }
        return true
    }

    private fun stepForces() {
        val pageList = _positions.keys.toList()
        val n = pageList.size
        if (n == 0) return

        val k  = cbrt((width * height * depth) / max(n.toDouble(), 1.0))
        val innerRadius = sphereRadius * (1 - SPHERE_MARGIN)

        val forces = HashMap<Page, DoubleArray>()
        for (p in pageList) forces[p] = DoubleArray(3)

        // Repulsion between all pairs
        for (i in 0 until n) {
            val u = pageList[i]; val pu = _positions[u]!!
            for (j in i + 1 until n) {
                val v = pageList[j]; val pv = _positions[v]!!
                val dx = pu.x - pv.x; val dy = pu.y - pv.y; val dz = pu.z - pv.z
                val dist = max(sqrt(dx * dx + dy * dy + dz * dz), 0.001)
                val rep = (k * k) / dist
                val fu = forces[u]!!; val fv = forces[v]!!
                fu[0] += (dx / dist) * rep; fu[1] += (dy / dist) * rep; fu[2] += (dz / dist) * rep
                fv[0] -= (dx / dist) * rep; fv[1] -= (dy / dist) * rep; fv[2] -= (dz / dist) * rep
            }
        }

        // Attraction along edges weighted by shared-link count
        for (u in pageList) {
            for (v in graph.linksFrom(u)) {
                if (!_positions.containsKey(v)) continue
                val pu = _positions[u]!!; val pv = _positions[v]!!
                val dx = pu.x - pv.x; val dy = pu.y - pv.y; val dz = pu.z - pv.z
                val dist = max(sqrt(dx * dx + dy * dy + dz * dz), 0.001)
                val att = (dist * dist) / k * (1.0 + graph.sharedLinkCount(u, v))
                val fu = forces[u]!!; val fv = forces[v]!!
                fu[0] -= (dx / dist) * att; fu[1] -= (dy / dist) * att; fu[2] -= (dz / dist) * att
                fv[0] += (dx / dist) * att; fv[1] += (dy / dist) * att; fv[2] += (dz / dist) * att
            }
        }

        // Spherical boundary repulsion: repel nodes that exceed the inner radius toward centre
        for (p in pageList) {
            if (p == pinnedPage) continue
            val pos = _positions[p]!!; val f = forces[p]!!
            val dx = pos.x - cx; val dy = pos.y - cy; val dz = pos.z - cz
            val dist = sqrt(dx * dx + dy * dy + dz * dz)
            if (dist > innerRadius && dist > 0.001) {
                val t = (dist - innerRadius) / (sphereRadius - innerRadius)
                val mag = k * t * t * BOUNDARY_STRENGTH
                f[0] -= (dx / dist) * mag
                f[1] -= (dy / dist) * mag
                f[2] -= (dz / dist) * mag
            }
        }

        // Integrate: skip frozen nodes entirely
        for (p in pageList) {
            if (p == pinnedPage) continue

            var count = stillCounts.getOrDefault(p, 0)
            if (count >= SETTLE_STEPS) continue

            val f = forces[p]!!; val vel = velocities[p]!!
            val pos = _positions[p]!!

            val pv0 = vel[0]; val pv1 = vel[1]; val pv2 = vel[2]

            vel[0] = (vel[0] + f[0] * FORCE_SCALE) * BASE_DAMPING
            vel[1] = (vel[1] + f[1] * FORCE_SCALE) * BASE_DAMPING
            vel[2] = (vel[2] + f[2] * FORCE_SCALE) * BASE_DAMPING

            var speed = sqrt(vel[0] * vel[0] + vel[1] * vel[1] + vel[2] * vel[2])
            if (speed > MAX_SPEED) {
                val s = MAX_SPEED / speed
                vel[0] *= s; vel[1] *= s; vel[2] *= s
                speed = MAX_SPEED
            }

            val dot = pv0 * vel[0] + pv1 * vel[1] + pv2 * vel[2]
            if (dot < 0) {
                vel[0] *= REVERSAL_DAMP; vel[1] *= REVERSAL_DAMP; vel[2] *= REVERSAL_DAMP
                speed  *= REVERSAL_DAMP
                count = min(count + REVERSAL_CREDIT, SETTLE_STEPS)
            } else if (speed < SETTLE_THRESHOLD) {
                count = min(count + 1, SETTLE_STEPS)
            } else {
                count = max(0, count - FAST_DECAY)
            }
            stillCounts[p] = count

            if (count >= SETTLE_STEPS) {
                vel[0] = 0.0; vel[1] = 0.0; vel[2] = 0.0
            } else if (count > 0) {
                val suppress = 1.0 - count.toDouble() / SETTLE_STEPS
                vel[0] *= suppress; vel[1] *= suppress; vel[2] *= suppress
            }

            pos.x = clamp(pos.x + vel[0], 0.0, width)
            pos.y = clamp(pos.y + vel[1], 0.0, height)
            pos.z = clamp(pos.z + vel[2], 0.0, depth)
        }

        // Clamp pinned page to center
        val pp = pinnedPage?.let { _positions[it] }
        if (pp != null) {
            pp.x = width  / 2.0
            pp.y = height / 2.0
            pp.z = 0.0
        }
    }

    fun step() {
        when (phase) {
            Phase.IDLE -> return
            Phase.ANIMATING -> {
                animStep++
                val t  = min(animStep.toDouble() / TRANSITION_STEPS, 1.0)
                val t2 = t * t * (3 - 2 * t)

                for ((page, pos) in _positions) {
                    val start  = animStart[page]  ?: continue
                    val target = animTarget[page] ?: continue
                    pos.x = start.x + (target.x - start.x) * t2
                    pos.y = start.y + (target.y - start.y) * t2
                    pos.z = start.z + (target.z - start.z) * t2
                }

                if (animStep >= TRANSITION_STEPS) {
                    for ((page, target) in animTarget) {
                        val pos = _positions[page] ?: continue
                        pos.x = target.x
                        pos.y = target.y
                        pos.z = target.z
                    }
                    phase = Phase.IDLE
                    animStart.clear()
                    animTarget.clear()
                }
            }
        }
    }

    private fun clamp(v: Double, lo: Double, hi: Double): Double = max(lo, min(hi, v))
}
